crates/matrix-sdk/src/notification_settings/mod.rs

```diff
@@ -352,6 +353,72 @@ impl NotificationSettings {
         }
     }
 
+    /// Get the keywords which have enabled rules.
+    pub async fn enabled_keywords(&self) -> IndexSet<String> {
+        self.rules.read().await.enabled_keywords()
+    }
+
+    /// Add or enable a rule for the given keyword.
```

Comment:
Seems like concurrent calls to this methods would trample over each other, right? Could you either change that or document it if changing it would be too problematic?

Reply:
I am not sure what the problem is. Afaict the request from this method is idempotent, so concurrent calls shouldn't cause issues?

Reply:
Or do you mean calling other methods from this API concurrently? Indeed it could create issues.

Then should we add a `Mutex` to make sure only one change is applied at a time?

Reply:
I think I get what you mean, hm. I just saw `.read().await.clone()` + a later `.write().await` and assumed that would immediately lead to problems. And indeed it sounds like there is potential for problems, just not as trivially as I thought.

A mutex sounds like a good idea, I think we even have a mechanism in the SDK already for request deduplication that handles a lot of edge cases. cc @bnjbvr

Reply:
Yeah, we could use the `DeduplicatingHandler` here, but we don't *have* to: the ctor overwrites the internal state with the latest push rules obtained from the server. Even if there are multiple concurrent calls to this method, the internal rules state will be eventually consistent (from the data read in the remote echo).

Reply:
Also the read-first-then-write latest pattern is quite safe here: we're only writing after effectful (`?`) statements :+1: